### ISSUE TITLE
Record/Table Tree overlay

### ISSUE THREAD

**Issue by spacejam** [spacejam/sled#562]
A nice feature for more structured queries, especially once transactions are in place on the Tree, is to be able to define a relatively thin Record or Table-like overlay where a rust structure is mapped into an ordering of types, some of which existing as a component of a primary key, and the rest being in the value. This will allow us to write code similar to:

```
#[derive(Record)]
struct Post {
  id: u64,
  author: String,
  content: String,
  author_ip: Ipv6Addr,
}

fn main() -> Result<(), ()> {
  let db = Db::start_default(path)?;
  let posts = db.open_table::<Post>(b"posts")?;
  
  posts.tx(|p| {
    p.update(|post| post.content = post.content.to_uppercase())
     .where(|post| post.author == "alice");
  })
}
```

**Comment by Kerollmops** [comment]
We could probably specify the primary key by specifying it using a container attribute [like serde do already](https://serde.rs/attributes.html).

```rust
#[derive(Record)]
#[sled(primary_key = id)]
struct Post {
  id: u64,
  author: String,
  content: String,
  author_ip: Ipv6Addr,
}
```

**Comment by spacejam** [comment]
@Kerollmops yeah! I was thinking this could also work for things like `auto_increment`, `unique`, `foreign_key(b"other_table", referenced_column_in_other_table)` that can be set on individual members.

**Comment by spacejam** [comment]
I don't want to get too deep into SQL territory but I consider the above things kind of table stakes ;)

**Comment by Kerollmops** [comment]
I think that the `Record` proc_macro should just be a convenient maner to implement a trait, this trait should implementable by hand, like serde `Serialize/Deserialize`. This way it could be possible to support dynamic types (e.g. json values, HashMaps).

**Comment by tomhoule** [comment]
I have been playing around with the idea for the prisma-sled connector I have been thinking about, but just noticed this issue, so I put together a very rough working sketch here: https://github.com/tomhoule/sled-record

The interesting part is mostly the test/example with a rountrip: https://github.com/tomhoule/sled-record/blob/master/sled-record/tests/generic_tests.rs

With field annotations, we could even support automatic secondary indexes (and `unique`), and generate methods on the struct, something like (if we had a unique index on `User.email`): `User::find_by_email(&db, "alice@example.com")`.

I'll try to be on discord tomorrow (berlin time) to discuss :)

**Comment by D1plo1d** [comment]
@tomhoule Very interesting!

I've put together a prototype of Sled record (de)serialization as well but with a focus on zero downtime migrations.

My thinking is that if we store records as enums that can be migrated on retrieval then an embedded application has a path forward for migrating legacy data. The specifics of the migration strategy are left to the application so this is very much just tools to (de)serialize forwards-compatible data to sled. Here's what it looks like in practice:

https://github.com/tegapp/teg/blob/bccbd75fcd7a2a52eb7308a025e2ff56fe73c800/packages/teg-auth/src/print_queue/tasks/models/mod.rs

One thing that interested me but I have zero experience in is indexes. How are you handling those in sled-record?
